tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
         // Less aggressive warnings to make the rustc toolstate management less painful.
         // (We often get warnings when e.g. a feature gets stabilized or some lint gets added/improved.)
         flags.push("-Astable-features".to_owned());
     } else {
-        flags.push("-Dwarnings -Dunused".to_owned()); // overwrite the -Aunused in compiletest-rs
+        flags.push("-Dwarnings".to_owned());
+        flags.push("-Dunused".to_owned()); // overwrite the -Aunused in compiletest-rs
     }
     if let Ok(sysroot) = env::var("MIRI_SYSROOT") {
-        flags.push(format!("--sysroot {}", sysroot));
+        flags.push("--sysroot".to_string());
+        flags.push(sysroot);
     }
     if let Ok(extra_flags) = env::var("MIRIFLAGS") {
-        flags.push(extra_flags);
+        for flag in extra_flags.split_whitespace() {
+            flags.push(flag.to_string());
+        }
     }
+    flags.push("-Zui-testing".to_string());
+    flags.push("--target".to_string());
+    flags.push(target.to_string());
+    let target = target.to_string();
 
-    let flags = flags.join(" ");
-    eprintln!("   Compiler flags: {}", flags);
+    eprintln!("   Compiler flags: {:?}", flags);
 
-    // The rest of the configuration.
-    let mut config = compiletest::Config::default().tempdir();
-    config.mode = mode.parse().expect("Invalid mode");
-    config.rustc_path = miri_path();
-    if let Some(lib_path) = option_env!("RUSTC_LIB_PATH") {
-        config.run_lib_path = PathBuf::from(lib_path);
-        config.compile_lib_path = PathBuf::from(lib_path);
-    }
-    config.filters = env::args().nth(1).into_iter().collect();
-    config.host = get_host();
-    config.src_base = PathBuf::from(path);
-    config.target = target.to_owned();
-    config.target_rustcflags = Some(flags);
-    compiletest::run_tests(&config);
-}
+    let grab_entries =
+        |path: &Path| std::fs::read_dir(path).unwrap().map(|entry| entry.unwrap().path());
+    let todo = SegQueue::new();
+    todo.push(PathBuf::from(path));
+
+    let failures = Mutex::new(vec![]);
+    let total = AtomicUsize::default();
+    let skipped = AtomicUsize::default();
+
+    crossbeam::scope(|s| {
+        for _ in 0..num_cpus::get() {
+            s.spawn(|_| {
+                while let Some(path) = todo.pop() {
+                    // Collect everything inside directories
+                    if path.is_dir() {
+                        for entry in grab_entries(&path) {
+                            todo.push(entry);
+                        }
+                        continue;
+                    }
+                    // Only look at .rs files
+                    if let Some(ext) = path.extension() {
+                        if ext != "rs" {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                    total.fetch_add(1, Ordering::Relaxed); // Read rules for skipping from file
```

Comment:
I don't understand the comment. Who is reading which rules here as part of the `fetch_add`?

Reply:
Huh... not sure what I meant

Reply:
oh it belongs to the next line 🤦